l1-contracts/contracts/bridge/L1Nullifier.sol

```diff
@@ -408,10 +398,20 @@ contract L1Nullifier is IL1Nullifier, ReentrancyGuard, Ownable2StepUpgradeable,
         isWithdrawalFinalized[chainId][l2BatchNumber][l2MessageIndex] = true;
 
         // Handling special case for withdrawal from ZKsync Era initiated before Shared Bridge.
-        require(!_isPreSharedBridgeEraEthWithdrawal(chainId, l2BatchNumber), "L1N: legacy eth withdrawal");
-        require(!_isPreSharedBridgeEraTokenWithdrawal(chainId, l2BatchNumber), "L1N: legacy token withdrawal");
+        (bytes32 assetId, bytes memory transferData, WithdrawalType withdrawalType) = _verifyWithdrawal(
+            _finalizeWithdrawalParams
+        );
+
+        // Handling special case for withdrawal from zkSync Era initiated before Shared Bridge.
+        if (withdrawalType == WithdrawalType.BaseToken && _isPreSharedBridgeEraEthWithdrawal(chainId, l2BatchNumber)) {
+            // Checks that the withdrawal wasn't finalized already.
+            bool alreadyFinalized = IGetters(ERA_DIAMOND_PROXY).isEthWithdrawalFinalized(l2BatchNumber, l2MessageIndex);
+            require(!alreadyFinalized, "Withdrawal is already finalized 2");
+        } else if (withdrawalType == WithdrawalType.LegacyTokenBridgeOrSharedBridge) {
+            require(!legacyBridge.isWithdrawalFinalized(l2BatchNumber, l2MessageIndex), "ShB: legacy withdrawal");
```

Comment:
this should be checked only for era chain id

Reply:
also, above you use `_isPreSharedBridgeEraEthWithdrawal`, but here you do not check that the withdrawal was an old one, we should choose what we do: either always check or not always check whether the withdraw was previously finalized